lib/python/picongpu/pypicongpu/species/util/element.py

```diff
@@ -7,65 +7,94 @@
 
 from ...rendering import RenderedObject
 
+import pydantic
 import typeguard
-import enum
+import typing
 import scipy
+import periodictable
+import re
 
 
 @typeguard.typechecked
-class Element(RenderedObject, enum.Enum):
+class Element(RenderedObject, pydantic.BaseModel):
     """
     Denotes an element from the periodic table of elements
 
     Used to provide fundamental constants for elements, and to map them in a
     type-safe way to PIConGPU.
 
-    The number associated is the number of protons.
-    Note: Spelling follows periodic table, e.g. "Na", "C", "He"
+    The number associated is just an id.
+    Note: Spelling follows periodic table, e.g. "Na", "C", "He" + typical nuclear variations
 
     Note that these denote Elements, but when initialized in a species *only*
-    represent the core, i.e. there are no electrons. To make an atom also
-    initialize an appropriate ionization.
+    describe the core, i.e. without electrons.
+    To describe atoms/ions you also need to initialize the charge_state of the species.
     """
 
-    H = 1
-    """hydrogen"""
-    He = 2
-    """helium"""
-    N = 7
-    """nitrogen"""
+    _store: typing.Optional[periodictable.core.Element] = None
 
     @staticmethod
-    def get_by_openpmd_name(openpmd_name: str) -> "Element":
+    def parse_openpmd_isotopes(openpmd_name: str) -> tuple[int | None, str]:
+        if openpmd_name == "":
+            raise ValueError(f"{openpmd_name} is not a valid openPMD particle type")
+        if openpmd_name[0] != "#" and re.match(r"[A-Z][a-z]?$|n$", openpmd_name):
+            return None, openpmd_name
+
+        m = re.match(r"#([1-9][0-9]*)([A-Z][a-z]?)$", openpmd_name)
+
+        if m is None:
+            raise ValueError(f"{openpmd_name} is not a valid openPMD particle type")
```

Comment:
What are valid openPMD particle types?

Reply:
as defined in [openPMD upcoming 2.0.0 particle type extension](https://github.com/openPMD/openPMD-standard/blob/upcoming-2.0.0/EXT_SpeciesType.md) under atoms and isotopes